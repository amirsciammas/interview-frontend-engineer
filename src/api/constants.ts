export const PRODUCTION_URL = 'https://jsonplaceholder.typicode.com';
