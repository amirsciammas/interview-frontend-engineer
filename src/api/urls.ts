const BASE_URL = "http://jsonplaceholder.typicode.com";
const USERS = `${BASE_URL}/users`

export { USERS }