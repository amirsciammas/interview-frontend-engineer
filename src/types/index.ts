export * from "./userTypes"
export * from "./response"