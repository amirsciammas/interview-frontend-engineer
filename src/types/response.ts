
export type Response<T> = {
    isError: boolean
    data: T
    message?: string
}