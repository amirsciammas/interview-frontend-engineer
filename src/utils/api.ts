import axios, { AxiosResponse } from 'axios';

export const BASE_URL = 'https://jsonplaceholder.typicode.com'
export const UserPostsEndpoints = {
  USERS: '/users',
  POSTS: '/posts'
}

function getResponseData<T extends object>(response: AxiosResponse<T>): T {
  return response.data;
}

export const getSimpleEndpoint = <T extends object>(endpoint: string) : Promise<T> => {
  return axios
  .get(BASE_URL + endpoint)
  .then((response) => getResponseData(response))
  .catch((error) => {
    console.log(error.response);
    throw error;
  })
}

export const getUsers = <T extends object>() : Promise<T> => {
  return getSimpleEndpoint(UserPostsEndpoints.USERS)
}


export const getUser = <T extends object>(userId: string) : Promise<T> => {
  return axios
  .get(`${BASE_URL + UserPostsEndpoints.USERS}/${userId}`)
  .then((response) => getResponseData(response))
  .catch((error) => {
    console.log(error.response);
    throw error;
  })
}

export const getPostsByUserId = <T extends object>(userId: string) : Promise<T> => {
  return axios
  .get(`${BASE_URL + UserPostsEndpoints.POSTS}/?userId=${userId}`)
  .then((response) => getResponseData(response))
  .catch((error) => {
    console.log(error.response);
    throw error;
  })
}
