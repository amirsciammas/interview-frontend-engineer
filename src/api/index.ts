import axios from "axios";
import { User } from "../types";
import { USERS } from "./urls"

const getUsers = () => axios.get<User>(USERS);


export { getUsers }