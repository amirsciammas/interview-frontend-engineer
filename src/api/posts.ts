import axios from 'axios';
import { Post } from '../types/post';
import to from '../utils/to';
import { PRODUCTION_URL } from './constants';

/**
 * Fetches the Post List data from JSONPlaceholder public API.
 */
 export const getPosts = async (): Promise<Array<Post>> => {
  const [error, response] = await to(
    axios.get(`${PRODUCTION_URL}/posts`)
  );
  if (error) {
    throw new Error(error);
  }
  return response.data;
};