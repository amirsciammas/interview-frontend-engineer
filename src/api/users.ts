import axios from 'axios';
import { User } from '../types/user';
import to from '../utils/to';
import { PRODUCTION_URL } from './constants';

/**
 * Fetches the User List data from JSONPlaceholder public API.
 */
 export const getUsers = async (): Promise<Array<User>> => {
  const [error, response] = await to(
    axios.get(`${PRODUCTION_URL}/users`)
  );
  if (error) {
    throw new Error(error);
  }

  return response.data;
};

/**
 * Fetches the User data by Id from JSONPlaceholder public API.
 */
 export const getUserById = async (userId: number): Promise<User> => {
  const [error, response] = await to(
    axios.get(`${PRODUCTION_URL}/users/${userId}`)
  );
  if (error) {
    throw new Error(error);
  }

  return response.data;
};