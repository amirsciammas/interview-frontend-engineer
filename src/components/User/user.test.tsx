import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { User } from 'src/types/api';
import UserComponent from './index';

function MockBrowserRouter({ children }: { children: React.ReactChild }) {
  return (
    <MemoryRouter initialEntries={['/']}>
      { children }
    </MemoryRouter>
  );
}

const mockUser: User = {
  id: 1,
  name: 'Leanne Graham',
  username: 'Bret',
  email: 'Sincere@april.biz',
  address: {
    street: 'Kulas Light',
    suite: 'Apt. 556',
    city: 'Gwenborough',
    zipcode: '92998-3874',
    geo: {
      lat: '-37.3159',
      lng: '81.1496',
    },
  },
  phone: '1-770-736-8031 x56442',
  website: 'hildegard.org',
  company: {
    name: 'Romaguera-Crona',
    catchPhrase: 'Multi-layered client-server neural-net',
    bs: 'harness real-time e-markets',
  },
};

test('user component renders correctly', () => {
  const component = renderer.create(
    <MockBrowserRouter><UserComponent user={mockUser} /></MockBrowserRouter>,
  );
  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});
