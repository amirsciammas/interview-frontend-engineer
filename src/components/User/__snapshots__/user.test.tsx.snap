// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user component renders correctly 1`] = `
<div
  style={
    Object {
      "margin": "0 auto 12px",
      "width": "70%",
    }
  }
>
  <div>
    <div>
      <div>
        <b>
          Bret
          ,
        </b>
         
        Romaguera-Crona
      </div>
      <div>
        <div>
          <p>
            <svg
              fill="none"
              height="1em"
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M15.05 5A5 5 0 0 1 19 8.95M15.05 1A9 9 0 0 1 23 8.94m-1 7.98v3a2 2 0 0 1-2.18 2 19.79 19.79 0 0 1-8.63-3.07 19.5 19.5 0 0 1-6-6 19.79 19.79 0 0 1-3.07-8.67A2 2 0 0 1 4.11 2h3a2 2 0 0 1 2 1.72 12.84 12.84 0 0 0 .7 2.81 2 2 0 0 1-.45 2.11L8.09 9.91a16 16 0 0 0 6 6l1.27-1.27a2 2 0 0 1 2.11-.45 12.84 12.84 0 0 0 2.81.7A2 2 0 0 1 22 16.92z"
              />
            </svg>
            1-770-736-8031 x56442
          </p>
          <p>
            <svg
              fill="none"
              height="1em"
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                r="10"
              />
              <line
                x1="2"
                x2="22"
                y1="12"
                y2="12"
              />
              <path
                d="M12 2a15.3 15.3 0 0 1 4 10 15.3 15.3 0 0 1-4 10 15.3 15.3 0 0 1-4-10 15.3 15.3 0 0 1 4-10z"
              />
            </svg>
            <a
              href="hildegard.org"
              rel="noopener noreferrer"
              target="_blank"
            >
              hildegard.org
            </a>
          </p>
          <p>
            <svg
              fill="none"
              height="1em"
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M4 4h16c1.1 0 2 .9 2 2v12c0 1.1-.9 2-2 2H4c-1.1 0-2-.9-2-2V6c0-1.1.9-2 2-2z"
              />
              <polyline
                points="22,6 12,13 2,6"
              />
            </svg>
            <a
              href="mailto:Sincere@april.biz"
            >
              Sincere@april.biz
            </a>
          </p>
        </div>
      </div>
    </div>
    <div>
      <a
        href="/users/1/posts"
        onClick={[Function]}
      >
        browse posts
      </a>
    </div>
  </div>
</div>
`;
